### PR TITLE
Add Gefen optimizer option

### PR DESCRIPTION
**Pull request by ndvbd**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

Adds Gefen as an optional optimizer choice in Megatron-LM.

## Issue tracking

Linked issue: Related to 5413

### Pre-checks

- [ ] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

## Details

This PR adds `gefen` as a supported value for the `--optimizer` argument.

When `--optimizer gefen` is selected, Megatron-LM instantiates `Gefen` with the same optimizer arguments used by Adam-style optimizers:

- `lr`
- `betas`
- `eps`
- `weight_decay`
- `fused=True`

Gefen remains optional. If `--optimizer gefen` is used but the Gefen package is unavailable, Megatron-LM raises an explicit import error.

## Testing

Tested locally with a small GPT mock-data training run using:

```bash
torchrun --nproc_per_node=1 pretrain_gpt.py \
  --num-layers 2 \
  --hidden-size 128 \
  --num-attention-heads 4 \
  --seq-length 128 \
  --max-position-embeddings 128 \
  --micro-batch-size 1 \
  --global-batch-size 1 \
  --train-iters 2 \
  --lr 1e-4 \
  --min-lr 1e-5 \
  --eval-interval 1000 \
  --eval-iters 1 \
  --lr-decay-style cosine \
  --lr-warmup-iters 1 \
  --weight-decay 0.01 \
  --tensor-model-parallel-size 1 \
  --pipeline-model-parallel-size 1 \
  --transformer-impl transformer_engine \
  --bf16 \
  --fp8-format hybrid \
  --fp8-amax-compute-algo max \
  --fp8-amax-history-len 16 \
  --mock-data \
  --optimizer gefen



### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by ndvbd** [comment]
@NVIDIA/mcore-oncall

**Comment by janEbert** [comment]
Hey @ndvbd, we'll get back to you on this. It's relatively certain that we cannot accept the PR in Megatron-LM due to (1) its dependence on external libraries, (2) as we don't accept new optimizer code here, and (3) new features in general need to be carefully considered before just merging them in to prevent feature creep.
I've asked folks from the [Emerging-Optimizers repo](https://github.com/NVIDIA-NeMo/Emerging-Optimizers) whether the optimizer could be contributed there and they're still reviewing, so please be patient with us for a week. Thank you!

On another matter, this raised a discussion among our team on how to enable an API for custom optimizer use, e.g., by adding arguments that could point to a Megatron-compatibly implemented optimizer in an external Python module with desired args/kwargs to pass. So this may be the preferred option, which would also support your use case.

**Comment by ndvbd** [comment]
@janEbert 

Thank you Jan for taking the time to look into this. Two updates from the last hours:

1. Throughput experiments on a pretraining run of a 31.6B-parameter model based on the Nemotron-3 architecture (Mamba + MoE), using Megatron Bridge (NVIDIA's Megatron-Core training infrastructure), showed a 6× throughput improvement when moving from the 𝗔𝗱𝗮𝗺𝗪 optimizer (25 s/step, 86 tokens/s) to the 𝗚𝗲𝗳𝗲𝗻 optimizer (4.2 s/step, 520 tokens/s). This translates into millions of dollars in potential training cost savings.

One reason for the speedup is that, with the Gefen optimizer, no CPU offloading was required, thereby eliminating CPU↔GPU communication overhead associated with optimizer state transfers.

The throughput experiment, which used a smaller configuration than the final training run, was conducted on 8 × H100 GPUs with 80 GiB of VRAM each. With the AdamW optimizer, approximately 385 GB of data had to be offloaded to CPU memory. In contrast, with the Gefen optimizer, no offloading was necessary, and all tensors resided entirely in GPU memory, utilizing a total of approximately 504 GiB of VRAM

2. A few hours ago a variant called Gefen-Muon was released as pip package: https://github.com/ndvbd/Gefen/
We took the same Gefen optimizer and added Muon orthagonalization into the first moment (only first moment is needed there), and you get the exact same performance as Muon with x4 less memory. We believe it is extremely significant, and if you can assist the community to place it in [Emerging-Optimizers repo](https://github.com/NVIDIA-NeMo/Emerging-Optimizers) that would be lovely.

Thanks for assisting!

**Comment by janEbert** [comment]
Closing, as corresponding issue #5413 was also closed. Discussion about the merge into [Emerging-Optimizers](https://github.com/NVIDIA-NeMo/Emerging-Optimizers) should continue over there.
